cmd/snap-bootstrap/cmd_initramfs_mounts_cvm.go

```diff
@@ -44,9 +49,175 @@ func (*genericCVMModel) Grade() asserts.ModelGrade {
+	imageManifestFile, err := osReadFile(imageManifestFilePath)
+	if err != nil {
+		return ImageManifest{}, err
+	}
+
+	var im ImageManifest
+	err = json.Unmarshal(imageManifestFile, &im)
+	if err != nil {
+		return ImageManifest{}, err
+	}
+
+	if len(im.Partitions) < 1 {
+		return ImageManifest{}, fmt.Errorf("Invalid manifest: root partition not specified.")
+	}
+
+	if im.Partitions[0].Overlay != "lowerdir" {
+		return ImageManifest{}, fmt.Errorf("Invalid manifest: expected first partition to be used as lowerdir, %s was found instead.", im.Partitions[0].Overlay)
+	}
+
+	return im, nil
+}
+
+// generateMountsFromManifest is used to parse a manifest file which contains information about which
+// partitions should be used to compose the rootfs of the system using overlayfs.
+//
+// Only a single overlayfs lowerdir and a single overlayfs upperdir are supported. For the lowerdir, a dm-verity
+// root hash can be supplied which will be used during mounting. The writable layer can be encrypted as in CVMv1.
+//
+// If a writable layer is not specified in the manifest, a tmpfs-based layer is mounted as the upperdir of the
+// overlayfs. This is relevant in ephemeral confidential VM scenarios where the confidentiality of the writable
+// data is achieved through hardware memory encryption and not disk encryption (the writable data/system state
+// should never touch the disk).
+func generateMountsFromManifest(im ImageManifest, disk disks.Disk) ([]partitionMount, error) {
+	foundReadOnlyPartition := ""
+	foundWritablePartition := ""
+
+	partitionMounts := []partitionMount{}
+
+	// Configure the overlay filesystems mounts from the manifest.
+	for _, p := range im.Partitions {
+		pm := partitionMount{
+			Opts: &systemdMountOptions{},
+		}
+
+		pm.GptLabel = p.GptLabel
+
+		// All detected partitions are mounted by default under /run/mnt/<GptLabel of partition>
+		pm.Where = filepath.Join(boot.InitramfsRunMntDir, p.GptLabel)
+
+		if p.Overlay == "lowerdir" {
+			// XXX: currently only one lower layer is permitted. The rest, if found, are ignored.
+			if foundReadOnlyPartition != "" {
+				continue
+			}
+			foundReadOnlyPartition = pm.GptLabel
+
+			// systemd-mount will run fsck by default when attempting to mount the partition and potentially corrupt it.
+			// This will cause dm-verity to fail when attempting to set up the dm-verity mount.
+			// fsck should be/is run by the encrypt-cloud-image tool prior to generating dm-verity data.
+			pm.Opts.NeedsFsck = false
+			pm.Opts.VerityRootHash = p.RootHash
+
+			// Auto-discover verity device from disk for partition types meant to be used as lowerdir
+			verityPartition, err := disk.FindMatchingPartitionWithPartLabel(p.GptLabel + "-verity")
+			if err != nil {
+				return []partitionMount{}, err
+			}
+			pm.Opts.VerityHashDevice = verityPartition.KernelDeviceNode
+		} else {
+			// Only one writable partition is permitted.
+			if foundWritablePartition != "" {
+				continue
```

Comment:
Same regarding log/error out

Reply:
See previous comment. This will now be stricter and throw an error.